src/m_argv.c

```diff
@@ -261,6 +261,127 @@ void M_FindResponseFile(void)
     }
 }
 
+void M_AddLooseFiles(void)
+{
+    int i, j, k;
+
+    char **wads;
+    char **lmps;
+    char **dehs;
+    int wadcount = 0;
+    int lmpcount = 0;
+    int dehcount = 0;
+
+    char **newargv;
+    int newargc;
+
+    struct {
+        const char *ext;
+        char ***list;
+        int *count;
+    } looses[] = {
+        {".wad", &wads, &wadcount},
+//      {".lmp", &wads, &wadcount}, // demo lumps are the most common
+        {".deh", &dehs, &dehcount},
+//      {".bex", &dehs, &dehcount},
+        {".hhe", &dehs, &dehcount},
+        {".seh", &dehs, &dehcount},
+        {".lmp", &lmps, &lmpcount},
+    };
+
+    struct {
+        const char *cmdparam;
+        char ***list;
+        int *count;
+    } params[] = {
+        {"-merge"   , &wads, &wadcount},
+        {"-deh"     , &dehs, &dehcount},
+        {"-playdemo", &lmps, &lmpcount},
+    };
+
+    // check for command line arguments
+    if (myargc < 2)
+    {
+        return;
+    }
+
+    // check for regular arguments and response files
+    for (i = 1; i < myargc; i++)
+    {
+        if (myargv[i][0] == '-' || myargv[i][0] == '@')
+        {
+            return;
+        }
+    }
+
+    // we will find at most myargc files of a kind
+    for (j = 0; j < arrlen(params); j++)
+    {
+        *(params[j].list) = malloc(myargc * sizeof(char *));
+    }
+
+    // identify loose files on the command line by their extensions
+    for (i = 1; i < myargc; i++)
+    {
+        char *lower;
+
+        lower = M_StringDuplicate(myargv[i]);
+        M_ForceLowercase(lower);
+
+        for (j = 0; j < arrlen(looses); j++)
+        {
+            if (M_StringEndsWith(lower, looses[j].ext))
+            {
+                (*(looses[j].list))[(*looses[j].count)++] = myargv[i];
+                break;
+            }
+        }
+
+        free(lower);
+    }
+
+    // count the new number of command line arguments,
+    // we need one additional regular parameter for each kind of file
+    newargc = 1;
+    for (j = 0; j < arrlen(params); j++)
+    {
+        if ((*params[j].count) > 0)
+        {
+            newargc += (*params[j].count) + 1;
+        }
+    }
+
+    // compose the new command line
+    if (newargc > 1)
```

Comment:
If arguments don't match any of the extensions, maybe make a catch-all default list they get put into instead. Then this will always be true. For example:
```
chocolate-doom doom2.wad asdf.txt deca.wad blah.bin batman.deh
```
becomes
```
chocolate-doom asdf.txt blah.bin -iwad doom2.wad -file deca.wad -deh batman.deh 
```

Reply:
Either this, or we simply ignore and drop them from the command line, or we abort the whole process because we have detected an obviously faulty case of drag-and-drop usage.